File: backend_py/llm_proxy.py
import json

import httpx
from fastapi.responses import StreamingResponse


def _error_event(message: str) -> bytes:
    return f"data: {json.dumps({'proxy_error': message})}\n\n".encode()


async def stream_chat(payload: dict) -> StreamingResponse:
    base_url = payload["base_url"].rstrip("/")
    headers = {"Content-Type": "application/json"}
    if payload.get("api_key"):
        headers["Authorization"] = f"Bearer {payload['api_key']}"

    body = {
        "model": payload["model"],
        "messages": payload["messages"],
        "stream": True,
    }
    if payload.get("tools"):
        body["tools"] = payload["tools"]
    for key in ("temperature", "max_tokens"):
        if payload.get(key) is not None:
            body[key] = payload[key]

    async def relay():
        try:
            async with httpx.AsyncClient(timeout=httpx.Timeout(300, connect=30)) as client:
                async with client.stream(
                    "POST", f"{base_url}/chat/completions", json=body, headers=headers
                ) as resp:
                    if resp.status_code != 200:
                        detail = (await resp.aread()).decode(errors="replace")
                        yield _error_event(f"Upstream {resp.status_code}: {detail[:2000]}")
                        return
                    async for chunk in resp.aiter_bytes():
                        yield chunk
        except httpx.HTTPError as e:
            yield _error_event(f"Connection error: {e}")

    return StreamingResponse(relay(), media_type="text/event-stream")

File: backend_js/llm_proxy.js
function errorEvent(message) {
  return `data: ${JSON.stringify({ proxy_error: message })}\n\n`;
}

export async function streamChat(payload, res) {
  const baseUrl = String(payload.base_url || '').replace(/\/+$/, '');
  const headers = { 'Content-Type': 'application/json' };
  if (payload.api_key) headers.Authorization = `Bearer ${payload.api_key}`;

  const body = {
    model: payload.model,
    messages: payload.messages,
    stream: true,
  };
  if (payload.tools) body.tools = payload.tools;
  for (const key of ['temperature', 'max_tokens']) {
    if (payload[key] !== undefined && payload[key] !== null) body[key] = payload[key];
  }

  res.writeHead(200, {
    'Content-Type': 'text/event-stream',
    'Cache-Control': 'no-cache',
    Connection: 'keep-alive',
  });

  // Abort upstream when the browser disconnects (Stop button) or after 300s overall.
  const controller = new AbortController();
  res.on('close', () => controller.abort());
  const timeout = setTimeout(() => controller.abort(), 300_000);

  try {
    const resp = await fetch(`${baseUrl}/chat/completions`, {
      method: 'POST',
      headers,
      body: JSON.stringify(body),
      signal: controller.signal,
    });
    if (resp.status !== 200) {
      const detail = await resp.text();
      res.end(errorEvent(`Upstream ${resp.status}: ${detail.slice(0, 2000)}`));
      return;
    }
    for await (const chunk of resp.body) {
      res.write(chunk);
    }
    res.end();
  } catch (e) {
    if (!res.writableEnded) {
      res.end(errorEvent(`Connection error: ${e.message || e}`));
    }
  } finally {
    clearTimeout(timeout);
  }
}
